### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] False positive when use assertion to infer array type

### ISSUE THREAD

**Issue by OpportunityLiu** [typescript-eslint/typescript-eslint#12268]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=LAKAZgrgdgxgLgSwPZQARiUgFASlQb1FVRhQGc5UBDAJxtQF5UBtAXQG5Qj0l6sAbAKaUEjVAAZ2qUQB5UARikIA1MpzditGgDoADhDIALLPlQAPAFwLUAX2pkC5q1AgBbAEaCaAGlQBPZzdPehscThBiAHpIgD0Afi4Inj4hETFJaVQ5RWlVdSTNOmYEVm0-MQAmcKjYhJBuGmEIGjQtcJtEyFhEFFR3WlwCblIoCmo6K1NLVBcPL18AmaCvWzYxNnDuDBThTKYM2QUlPI1xnX0jEydrUOrUaPjE4m3UAV3RfaUso9y1U61iqVykwqtwHnUGk0Wmd2p1oPBkGh%2BgAvQaEJIjMZaSbXWbBBaBOYhNZMDZPZKvVJ7CRfbLHP4FM56AzGKZWeS2eyOaZ4%2Bb%2BQnBWxhMG1ckvN5pT6ZOm-fLEQo0QFlSp3cHkxpwZqtOiwkBAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYASSSomyPAEEiATwphR6KQF8Q%2BoA&tokens=false

### Repro Code

```TypeScript
function foo() {
  const arr = [];

  for (let i = 0; i < 1; i++)
    arr.push({ x: 1 } as { x: number, y: number });
  //^?

  for (let i = 0; i < 1; i++)
    arr[i].y = 2;
  //^?

  return arr;
}

function bar() {
  const arr: { x: number, y: number }[] = [];

  for (let i = 0; i < 1; i++)
    arr.push({ x: 1 });
  //^?

  for (let i = 0; i < 1; i++)
    arr[i].y = 2;
  //^?

  return arr;
}

function baz() {
  const arr: { x: number, y: number }[] = [];

  for (let i = 0; i < 1; i++)
    arr.push({ x: 1 } as { x: number, y: number });
  //^?

  for (let i = 0; i < 1; i++)
    arr[i].y = 2;
  //^?

  return arr;
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true,
    "noImplicitAny": true
  }
}
```

### Expected Result

Type assertion when push to an implicit `any[]` should be allowed to make typescript to infer the array type, as the `foo` method.

### Actual Result

Have to write the type twice as `baz` method.

**Comment by bradzacher** [comment]
I would say that relying on an evolving `any` is usually a bad idea for safety and generally it should be avoided and this case should be an annotation on the original variable instead of an assertion on the object.

That aside though - the rule should probably understand that an evolving `any` doesn't strictly mean `any`. I believe there are flags set on the type that would allow us to tell this.

But if we can't tell it directly then this would have to be closed as a wontfix - will require some investigation.
